aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,1141 @@
+            if (vote) {
+                stats.total_yes = stats.total_yes + 1;
+                bucket_table::add(&mut stats.yes_votes, token_id, voter_addr);
+            } else {
+                stats.total_no = stats.total_no + 1;
+                bucket_table::add(&mut stats.no_votes, token_id, voter_addr);
+            };
+            i = i + 1;
+        };
+
+        nft_dao_events::emit_voting_event(
+            voter_addr,
+            nft_dao,
+            proposal_id,
+            vote,
+            token_names,
+            property_versions,
+        );
+    }
+
+    /// Anyone can call the resolve function to resolve a proposal.
+    public entry fun resolve<CoinType>(proposal_id: u64, nft_dao: address) acquires Proposals, DAO, ProposalVotingStatistics {
+        let dao = borrow_global<DAO>(nft_dao);
+        // assert the proposal voting ended
+        let proposals = borrow_global<Proposals>(nft_dao);
+        let proposal = table::borrow(&proposals.proposals, proposal_id);
+        let now = timestamp::now_seconds();
+        assert!(now >= proposal.start_time_sec + dao.voting_duration, error::invalid_argument(EPROPOSAL_ENDED));
+        // assert the proposal is unresolved yet
+        assert!(proposal.resolution == PROPOSAL_PENDING, error::invalid_argument(EPROPOSAL_RESOLVED));
+        resolve_internal<CoinType>(option::none(), proposal_id, nft_dao);
+    }
+
+    /// Admin can veto an active proposal
+    public entry fun admin_veto_proposal(admin: &signer, proposal_id: u64, nft_dao: address) acquires DAO, Proposals {
+        let dao = borrow_global_mut<DAO>(nft_dao);
+        assert!(dao.admin == signer::address_of(admin), error::permission_denied(EINVALID_ADMIN_ACCOUNT));
+        // assert the proposal is still active
+        let proposals = borrow_global_mut<Proposals>(nft_dao);
+        let proposal = table::borrow_mut(&mut proposals.proposals, proposal_id);
+        // assert the proposal is unresolved yet
+        assert!(proposal.resolution == PROPOSAL_PENDING, error::invalid_argument(EPROPOSAL_RESOLVED));
+        proposal.resolution = PROPOSAL_VETOED_BY_ADMIN;
+
+        nft_dao_events::emit_resolve_event(
+            proposal_id,
+            nft_dao,
+            proposal.resolution,
+        )
+    }
+
+    /// DAO admin can directly resolve a proposal
+    public entry fun admin_resolve<CoinType>(admin: &signer, proposal_id: u64, nft_dao: address) acquires DAO, Proposals, ProposalVotingStatistics {
+        let resolver = signer::address_of(admin);
+        // assert the proposal voting ended
+        let proposals = borrow_global<Proposals>(nft_dao);
+        let proposal = table::borrow(&proposals.proposals, proposal_id);
+        // assert the proposal is unresolved yet
+        assert!(proposal.resolution == PROPOSAL_PENDING, error::invalid_argument(EPROPOSAL_RESOLVED));
+
+        resolve_internal<CoinType>(option::some(resolver), proposal_id, nft_dao);
```

Comment:
If we emit the same event for both admin resolve and normal resolve, we wouldn't be able to tell these 2 apart? May be we have a separate event for admin resolve or if we're using the same event, may be a field to indicate if it was an admin who did the resolution.

Reply:
similarly, added new events